modules/frame-nonce-preview.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/**
+ * Allows viewing posts on the frontend when the user is not logged in.
+ */
+class Jetpack_Frame_Nonce_Preview {
+	static $instance = null;
+
+	/**
+	 * Returns the single instance of the Jetpack_Frame_Nonce_Preview object
+	 *
+	 * @since 4.4.0
+	 *
+	 * @return Jetpack_Frame_Nonce_Preview
+	 **/
+	public static function get_instance() {
+		if ( ! is_null( self::$instance ) ) {
+			return self::$instance;
+		}
+
+		return self::$instance = new Jetpack_Frame_Nonce_Preview();
+	}
+
+	function __construct() {
+		if (
+			isset( $_GET['frame-nonce'] ) &&
+			! is_admin() &&
+			! is_user_logged_in()
+
+		) {
+			add_filter( 'pre_get_posts', array( $this, 'maybe_display_post' ) );
+		}
+	}
+
+	/**
+	 * Verify that frame nonce exists, and if so, validate the nonce by calling WP.com.
+	 *
+	 * @since 4.4.0
+	 *
+	 * @return bool
+	 */
+	public function is_frame_nonce_valid() {
+		if ( empty( $_GET[ 'frame-nonce' ] ) ) {
```

Comment:
Can we rename this into something like 'jp-preview-frame-nonce'
just so we avoid potential conflicts with some other plugin down the line. :)